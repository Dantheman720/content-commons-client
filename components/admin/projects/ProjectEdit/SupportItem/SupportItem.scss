.support-item {
  color: #000;
  text-shadow: none;

  &.new {
    color: rgba(192, 192, 192, 0.7);
    transition: 0.4s;

    .inline.loader {
      margin-left: 0.5rem;
    }
  }

  > .ui.progress {
    border-radius: 0;

    &:first-child {
      margin-top: auto;
      margin-bottom: auto;
      width: 60%;

      > .bar {
        border-radius: 0;

        &:after {
          border-radius: 0;
        }
      }
    }
  }

  > .item-lang {
      &.error {
      overflow: visible !important;
    }
  }
}
