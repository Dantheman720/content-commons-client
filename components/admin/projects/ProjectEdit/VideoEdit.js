import React from 'react';

const VideoEdit = () => (
  <div>Video Edit</div>
);

export default VideoEdit;
