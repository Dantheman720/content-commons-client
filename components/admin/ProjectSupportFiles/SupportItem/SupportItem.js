/**
 *
 * SupportItem
 *
 */
import React, {
  Fragment, useState, useEffect, useRef, useContext
} from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { compose, graphql } from 'react-apollo';
import { Loader, Popup } from 'semantic-ui-react';
import debounce from 'lodash/debounce';
import Focusable from 'components/Focusable/Focusable';
import GeneralError from 'components/errors/GeneralError/GeneralError';
import VisuallyHidden from 'components/VisuallyHidden/VisuallyHidden';
// import FileRemoveReplaceButtonGroup from 'components/admin/FileRemoveReplaceButtonGroup/FileRemoveReplaceButtonGroup';
import { LANGUAGES_QUERY } from 'components/admin/dropdowns/LanguageDropdown';
import { getPathToS3Bucket } from 'lib/utils';
import { UploadContext } from '../../ProjectEdit/VideoEdit/VideoEdit';

import './SupportItem.scss';

/* eslint-disable react/prefer-stateless-function */
const SupportItem = props => {
  const STR_INDEX_PROPORTION = 0.04;
  const ITEM_NAME_PROPORTION = 0.625;
  const ITEM_LANG_PROPORTION = 0.3;
  const DELAY_INTERVAL = 1000;

  const listEl = useRef( null );
  const filenameEl = useRef( null );
  const languageEl = useRef( null );
  const mounted = useRef( true );

  const uploadInProgress = useContext( UploadContext );

  const [error, setError] = useState( false );
  const [widths, setWidths] = useState( {
    listItem: 0,
    itemName: 0,
    itemLang: 0
  } );

  const { item } = props;
  const { listItem, itemName, itemLang } = widths;

  const updateWidths = () => {
    if ( mounted.current ) {
      setWidths( {
        listItem: listEl.current.offsetWidth,
        itemName: filenameEl.current.offsetWidth,
        itemLang: ( languageEl.current && languageEl.current.offsetWidth ) || 0
      } );
    }
  };

  const debounceResize = debounce( updateWidths, DELAY_INTERVAL );

  const checkFileUrlStatus = () => {
    if ( item && item.url ) {
      const { url } = item;
      const path = url;
      const options = { baseURL: getPathToS3Bucket() };
      axios.get( path, options )
        .catch( err => {
          console.dir( err );
          setError( err.isAxiosError );
        } );
    }
  };

  const hasError = () => {
    if ( item && typeof item.error !== 'undefined' ) {
      setError( item.error );
      return;
    }
    checkFileUrlStatus();
  };

  useEffect( () => {
    window.addEventListener( 'resize', debounceResize );
    updateWidths();

    return () => {
      window.removeEventListener( 'resize', debounceResize );
      mounted.current = false;
    };
  }, [] );

  useEffect( () => {
    hasError();
  }, [item.url] );

  /**
   * Truncates long strings with ellipsis
   * @param {string} str the string
   * @param {number} start index for first cutoff point
   * @param {number} end index for ending cutoff point
   * @return truncated string
   */
  const getShortFileName = ( str, index ) => (
    <Fragment>
      { `${str.substr( 0, index )}` }&hellip;{ `${str.substr( -index )}` }
    </Fragment>
  );

  /**
   * Determines an integer proportional
   * to a reference number
   * @param {number} reference
   * @return {number}
   */
  const getProportionalNumber = ( reference, proportion ) => (
    Math.floor( reference * proportion )
  );

  const isLongName = ( itemWidth, reference, proportion ) => (
    itemWidth >= getProportionalNumber( reference, proportion )
  );


  const normalizeLanguage = lang => {
    if ( typeof ( lang ) === 'string' ) { // will be object if coming from graphql
      return props.data.languages.filter( l => l.id === lang )[0];
    }
    return lang;
  };

  const normalizeItem = ( prop, defaultValue ) => {
    const fileProp = item[prop];

    if ( fileProp ) {
      return fileProp;
    } if ( item.input ) {
      return item.input[prop.replace( 'file', '' )];
    }
    return defaultValue;
  };

  const filename = normalizeItem( 'filename', '' );
  const filesize = normalizeItem( 'filesize', 0 );
  const language = normalizeLanguage( item.language );
  const charIndex = getProportionalNumber( listItem, STR_INDEX_PROPORTION );
  const shortFileName = getShortFileName( filename, charIndex );
  const isLongFileName = isLongName( itemName, listItem, ITEM_NAME_PROPORTION );
  const isLongLangName = isLongName( itemLang, listItem, ITEM_LANG_PROPORTION );
  const isUploading = uploadInProgress && ( item.loaded < filesize );

  const popupStyle = {
    overflowWrap: 'break-word',
    wordWrap: 'break-word',
    msWordBreak: 'break-all',
    wordBreak: 'break-word'
  };

  const renderName = ( str = '', isLang = false ) => {
    if ( isLongFileName ) {
      return (
        <Popup
          content={ str }
          size="mini"
          inverted
          on={ [
            'hover',
            'click',
            'focus'
          ] }
          trigger={ (
            <span>
              <Focusable>
                { isLang ? str : shortFileName }
              </Focusable>
            </span>
            ) }
          style={ popupStyle }
        />
      );
    }
    return str;
  };

  return (
    <li
      key={ `${item.id}-${language.id}` }
      className={ `support-item ${( item.loaded < filesize ) ? 'new' : ''}` }
      ref={ listEl }
    >
      <span className="item-name" style={ error ? { color: '#cd2026' } : {} }>
        { isLongFileName && <VisuallyHidden>{ filename }</VisuallyHidden> }
        <span
          className={
            `item-name-wrap${isLongFileName ? ' hasEllipsis' : ''}`
          }
          aria-hidden={ isLongFileName }
          ref={ filenameEl }
        >
          { error
            ? <GeneralError msg={ renderName( filename ) } />
            : renderName( filename ) }
        </span>

        <Loader active={ isUploading } inline size="mini" style={ { marginLeft: '.5rem' } } />

      </span>

      <span className={ `item-lang${error ? ' error' : ''}` }>
        { error
          ? (
            null
            /* <FileRemoveReplaceButtonGroup
              onReplace={ () => {} }
              onRemove={ () => console.log( 'removed' ) }
            /> */
          )
          : (
            <b
              className={
                `item-lang-wrap${isLongLangName ? ' hasEllipsis' : ''}`
              }
              ref={ languageEl }
            >
              { renderName( language.displayName, true ) }
            </b>
          ) }
      </span>
    </li>
  );
};


SupportItem.propTypes = {
  item: PropTypes.object,
  data: PropTypes.object
};

export default compose(
  graphql( LANGUAGES_QUERY )
)( SupportItem );
