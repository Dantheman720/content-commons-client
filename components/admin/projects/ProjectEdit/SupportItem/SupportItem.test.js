import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { projects, supportFilesConfig } from 'components/admin/projects/ProjectEdit/mockData';
import SupportItem from './SupportItem';

const props = {
  supportItem: projects[0].supportFiles.srt[0],
  projectId: { videoID: 'made-in-america' },
  fileType: supportFilesConfig.srt.fileType,
  itemId: projects[0].supportFiles.srt[0].id
};

const Component = <SupportItem { ...props } />;

describe( '<SupportItem />', () => {
  it( 'renders without crashing', () => {
    const wrapper = shallow( Component );
    expect( wrapper.exists() ).toEqual( true );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  it( 'renders `Progress` component if `isUploading`', () => {
    const wrapper = shallow( Component );

    expect( wrapper.state( 'isUploading' ) ).toEqual( false );
    expect( wrapper.find( 'Progress' ).exists() )
      .toEqual( wrapper.state( 'isUploading' ) );
    wrapper.setState( { isUploading: true } );
    expect( wrapper.state( 'isUploading' ) )
      .toEqual( wrapper.state( 'isUploading' ) );
    expect( wrapper.find( 'Progress' ).exists() )
      .toEqual( wrapper.state( 'isUploading' ) );
    expect( wrapper.find( 'Progress' ).prop( 'total' ) )
      .toEqual( props.supportItem.size.filesize );
    expect( toJSON( wrapper ) ).toMatchSnapshot();
  } );

  // it( 'renders error icon and message component if `supportItem` has an error `uploadStatus`', () => {
  //   const wrapper = shallow( Component );

  //   expect( wrapper.find( 'li.support-item.error' ).exists() )
  //     .toEqual( false );
  //   wrapper.setProps( {
  //     supportItem: {
  //       id: '5678',
  //       lang: 'Arabic',
  //       file: 'madeinamerica_arabic.srt',
  //       uploadStatus: {
  //         error: true,
  //         success: false
  //       },
  //       size: { filesize: 24576 }
  //     }
  //   } );
  //   expect( wrapper.find( 'li.support-item.error' ).exists() )
  //     .toEqual( true );
  //   expect( toJSON( wrapper ) ).toMatchSnapshot();
  // } );
} );
