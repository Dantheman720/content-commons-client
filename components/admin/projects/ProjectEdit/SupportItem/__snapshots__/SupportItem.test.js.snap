// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SupportItem /> renders \`Progress\` component if \`isUploading\` 1`] = `
<li
  className="support-item isUploading"
  key="srt-Arabic"
>
  <Progress
    active={true}
    color="blue"
    precision={0}
    progress={true}
    size="small"
    total={24576}
    value={0}
  />
  <span>
    Uploading item
  </span>
</li>
`;

exports[`<SupportItem /> renders error icon and message component if \`supportItem\` has an error \`uploadStatus\` 1`] = `
<li
  className="support-item error"
  key="srt-Arabic"
>
  <p>
    <Icon
      as="i"
      color="red"
      name="exclamation triangle"
      size="small"
    />
    <span>
      Uploading  error
    </span>
  </p>
</li>
`;

exports[`<SupportItem /> renders without crashing 1`] = `
<li
  className="support-item"
  key="srt-Arabic"
>
  <span
    className="item-name"
  >
    <VisuallyHidden>
      madeinamerica_arabic.srt
    </VisuallyHidden>
    <span
      aria-hidden={true}
      className="item-name-wrap hasEllipsis"
    >
      <Popup
        content="madeinamerica_arabic.srt"
        inverted={true}
        keepInViewPort={true}
        on={
          Array [
            "hover",
            "click",
            "focus",
          ]
        }
        position="top left"
        size="small"
        style={
          Object {
            "msWordBreak": "break-all",
            "overflowWrap": "break-word",
            "wordBreak": "break-word",
            "wordWrap": "break-word",
          }
        }
        trigger={
          <span>
            <Focusable
              order={0}
            >
              <React.Fragment>
                
                â€¦
                madeinamerica_arabic.srt
              </React.Fragment>
            </Focusable>
          </span>
        }
      />
    </span>
  </span>
  <span
    className="item-lang"
  >
    <b
      className="item-lang-wrap hasEllipsis"
    >
      <Popup
        content="Arabic"
        inverted={true}
        keepInViewPort={true}
        on={
          Array [
            "hover",
            "click",
            "focus",
          ]
        }
        position="top left"
        size="small"
        style={
          Object {
            "msWordBreak": "break-all",
            "overflowWrap": "break-word",
            "wordBreak": "break-word",
            "wordWrap": "break-word",
          }
        }
        trigger={
          <span>
            <Focusable
              order={0}
            >
              Arabic
            </Focusable>
          </span>
        }
      />
    </b>
  </span>
</li>
`;
